### PR TITLE
Use `appraisal` gem in CI

### PR DESCRIPTION
**Pull request by boardfish**
Supersedes #1230

Uses `appraisal ~> 2.5` to run against an array of Rails minor versions from `6.1` onwards. Also moves us away from manually caching gem dependencies, as that can now be handled by the `setup-ruby` action.

### CODE REVIEW

**Comment by Spone** [comment]
Looking good, thanks for wrapping this up! :tada: 

While re-reading #1230, I noticed there was one last thing to decide. Do we want to:
- merge this as it is (with the dependency on `excid3/appraisal` fork ([see this for background](https://github.com/github/view_component/pull/1230#discussion_r806892525))
- wait for https://github.com/thoughtbot/appraisal/pull/174 to be merged
- refactor this PR to not use appraisal directly in CI, but setting `BUNDLE_GEMFILE` via `matrix`, as suggested [here](https://github.com/github/view_component/pull/1230#discussion_r806926293)

Happy to hear what you think @joelhawksley @boardfish @BlakeWilliams

**Comment by boardfish** [comment]
I'm leaning towards that first option. I think we should prioritize ensuring that running the tests works the same way in dev as it does in CI. It may be a good idea in the long run to mirror @excid3's fork to @github so we can ensure that version of the code remains publicly available until @excid3's PR is merged.

**Review by camertron — APPROVED** [review]
Nice this is great! I'm a big fan of appraisal :)

**Comment by Spone** [comment]
@joelhawksley would you mind giving us your feedback on this? (See my latest comment above). I think we should try to merge this soon.

**Review by joelhawksley — CHANGES_REQUESTED** [review]
@Spone IMO we need to be cautious of depending on a long-running upstream PR on Appraisal.

Let's take another approach, or work to get the upstream PR merged.

**Comment by allmarkedup** [comment]
FWIW, I've just gone through the fun process of trying to get Lookbook CI working with Appraisal - I took a slightly different approach (and with a smaller matrix of ruby/rails versions for now) but I finally got it working. Just in case it's useful to have an example which uses the matrix approach but doesn't require the Appraisal fork: https://github.com/allmarkedup/lookbook/blob/main/.github/workflows/ci.yml

**Comment by Spone** [comment]
There is [a new PR](https://github.com/thoughtbot/appraisal/pull/202) on Appraisal that could fix the issue and allow us to move forward with this approach.

If it does not, I'll take some time to mirror @allmarkedup's method.

**Comment by nickcharlton** [comment]
Appraisal maintainer here! 👋 

I've just merged that PR mentioned above (https://github.com/thoughtbot/appraisal/pull/202), which I'm hoping can help you out here. In a few days I'll cut a new release.

We'd been a bit stuck with a good solution to that problem, but let us know over there if you're having problems.

**Comment by Spone** [comment]
Awesome @nickcharlton! I'll try it out next week. Thanks for the heads up!

**Comment by Spone** [comment]
Awesome @nickcharlton! I'll try it out next week. Thanks for the heads up!

File: Gemfile
Comment (Spone) [inline Gemfile:21]: I'm targeting `main` for now, but we'll be able to use the [next version](https://github.com/thoughtbot/appraisal/issues/204) when it's released.

**Comment by Spone** [comment]
I thought it was time to get this in shape for v3.0.0 :)

Do not forget this after merging: https://github.com/ViewComponent/view_component/pull/1308#discussion_r834585238

Thanks a lot @boardfish for your help on this!

**Comment by joelhawksley** [comment]
@Spone I'd prefer we wait for the Appraisal release instead of depending on a branch.

**Comment by joelhawksley** [comment]
@boardfish thoughts on testing this against https://github.com/thoughtbot/appraisal/issues/204#issuecomment-1635841844? Perhaps we don't need to wait for 3.0?

**Comment by boardfish** [comment]
Yeah, that sounds good!

**Comment by joelhawksley** [comment]
@boardfish cool, would you be up for bringing this PR up to date?

**Comment by boardfish** [comment]
Sure! I can have a crack at that sometime in the next week.

**Comment by boardfish** [comment]
Alright, I _think_ we're just about there. Personal TODO list of stuff I've recognized that's still left to do on this one:

- [ ] Check that the CC patch is still actually applied
- [ ] Figure out why simplecov's suggesting that `test_helpers.rb` isn't covered
- [ ] followup: see if we can use built-in bundler caching for other test runs instead of manually calling `actions/cache` to remove that layer of complexity from these workflows

File: .github/workflows/ci.yml
Comment (boardfish) [inline .github/workflows/ci.yml:75]: I'll restore this comment.

File: .github/workflows/ci.yml
Comment (boardfish) [inline .github/workflows/ci.yml:35]: Is there a reason we were previously selective about which versions we'd run against? Maybe it was a holdover from when we needed specific version combos. That would still be possible [this way](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs#expanding-or-adding-matrix-configurations).

File: test/test_helper.rb
Comment (boardfish) [inline test/test_helper.rb:4]: This is a holdover from when I had to deal with warnings being raised, which I think might've followed from the rebase, but it's unnecessary now. I'll remove it.

```suggestion
```

File: .github/workflows/ci.yml
Comment (camertron) [reply .github/workflows/ci.yml:35]: Yeah, certain versions of Rails require certain versions of Ruby.

**Review by camertron — COMMENTED** [review]
LGTM! Can you merge main?

File: .github/workflows/ci.yml
Comment (boardfish) [reply .github/workflows/ci.yml:35]: I was wondering if we'd still need to do this, but looking at [endoflife.date](https://endoflife.date/rails), the lists of versions we have here look good.

**Comment by camertron** [comment]
Ugh sorry @boardfish, could you fix these merge conflicts? I'm not sure how to resolve the difference in the Gemfile 😓

File: gemfiles/rails_main.gemfile
Comment (joelhawksley) [inline gemfiles/rails_main.gemfile:22]: Can we use the rubygems source here?

File: lib/view_component/test_helpers.rb
Comment (joelhawksley) [inline lib/view_component/test_helpers.rb:267]: Why were these changes necessary?

File: .github/workflows/ci.yml
Comment (camertron) [inline .github/workflows/ci.yml:72]:
```suggestion
        ENABLE_RELOADING=true bundle exec appraisal rails-${{ matrix.rails_version }} rake
```

File: .github/workflows/ci.yml
Comment (camertron) [inline .github/workflows/ci.yml:65]: 😎

**Review by camertron — CHANGES_REQUESTED** [review]

File: lib/view_component/test_helpers.rb
Comment (boardfish) [reply lib/view_component/test_helpers.rb:267]: I'll try dropping them, but runs were failing due to coverage errors. I think this was the cause, though I'm not sure if anything in this PR in particular would've triggered it.

**Comment by boardfish** [comment]
@joelhawksley Looks like we do still get [coverage-related failures](https://github.com/ViewComponent/view_component/actions/runs/7407617837/job/20154220996) without those nocov comments. Not sure why, though – maybe there's a discrepancy in the running environment between the runs on `main` vs. on this branch.

**Comment by joelhawksley** [comment]
@boardfish I'm happy to merge and revisit given the complexity of this PR.

**Comment by boardfish** [comment]
Awesome. I'll get rid of those merge conflicts and merge this in myself later, then 👍

**Pull request merged**
